### PR TITLE
[multistage] UNION/INTERSECT/EXCEPT implementation 

### PR DESCRIPTION
**Pull request by xiangfu0**
UNION/INTERSECT/EXCEPT(MINUS) implementation based on https://github.com/apache/pinot/pull/10535

For INTERSECT/EXCEPT(MINUS) implementation, the code assumes the right side is **SMALLER**, so it will go through the entire right side to construct a TreeSet data structure, then go through left side to decide what to put in the final results.

**DISCLAIMER**: Multi-value distinct is not supported, so `UNION` won't dedup on array columns. Will have a follow up PR to fix this issue. Follow up issue: https://github.com/apache/pinot/issues/10658

Union All example:
`select * from billing where city = 'Palo Alto' UNION ALL select * from billing where city = 'Mountain View'`

image

The result matches the two individual results (287 = 140 + 147):

`select * from billing where city = 'Palo Alto'`
image

`select * from billing where city = 'Mountain View'`
image

For Union example:

The dedup happens as expected for:
`select * from billing where city = 'Palo Alto' UNION select * from billing where city = 'Palo Alto'`
Scanned 280 rows and results is 140.
image

The result is the same as `select * from billing where city = 'Palo Alto'`

For Intersect example:
`select * from billing INTERSECT select * from billing where city = 'Mountain View'`
image

For Except(Minus) example:
`select * from billing EXCEPT select * from billing where city = 'Mountain View'`
image

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java [hunk 1]
```diff
@@ -105,8 +110,24 @@ public MultiStageOperator visitWindow(WindowNode node, PlanRequestContext contex
...
-        "Stage node of type SetOpNode: " + setOpNode.getSetOpType() + " is not supported!");
+    List<MultiStageOperator> inputs = new ArrayList<>();
+    for (StageNode input : setOpNode.getInputs()) {
+      MultiStageOperator visited = input.visit(this, context);
+      inputs.add(visited);
+    }
+    switch (setOpNode.getSetOpType()) {
+      case UNION:
+        return new UnionOperator(context.getOpChainExecutionContext(), inputs,
+            setOpNode.getInputs().get(0).getDataSchema());
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java:121]: this and others below: seems like it is assuming it always gets Union "all=true"?
do we plan to support all=false going forward? if so can we add check on `setOp.all` ? and also add a ignored test in SetOp.json test cases?

Comment (xiangfu0) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java:121]: For now UNION is parsed to UNION ALL + DISTINCT, so for UNION, all is always true.
For INTERSECT and MINUS, the all is always false.

We can support it the check for all later if parser side can support it.

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSetOpExchangeNodeInsertRule.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    return false;
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    SetOp setOp = call.rel(0);
+    List<RelNode> newInputs = new ArrayList<>();
+    for (RelNode input : setOp.getInputs()) {
+      RelNode exchange = LogicalExchange.create(input, RelDistributions.hash(ImmutableList.of(0)));
```
[the remark this answers is not in the document]
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSetOpExchangeNodeInsertRule.java]: Changed to hash all the columns.

**Review by walterddr — COMMENTED** [review]
lgtm overall. please fix the explain plan test

**Comment by xiangfu0** [comment]
> lgtm overall. please fix the explain plan test

done

**Review by walterddr — APPROVED** [review]

**Pull request merged**